man/POSIX/qsub.1
.SH "NAME"
.TH "qsub" 1
qsub \- submit jobs to the queue manager for execution
.SH "SYNOPSIS"
.B qsub 
.I [<options>] <executable> [<executable options>]
.SH "DESCRIPTION"
.PP
.B qsub
submits a job to the queue manager for execution. 

.B JOBID EXPANSION ($jobid or $COBALT_JOBID):

If $jobid or $COBALT_JOBID is specified in any jobid expansion option 
then it will be converted to the actual jobid. 

Example: qsub ... --env myenv=\\$jobid_myval ... 

If the jobid is 123 then myenv will be set to '123_myval'

jobid expansion options: --env, --jobname, -o, -e, -O, --debuglog

Note: Remember to escape appropriately "\\$jobid" or "\\$COBALT_JOBID" 

.B SCRIPT JOB DIRECTIVES (#COBALT):

Command options can be entered in a script job by using #COBALT <options>.
You can enter all in one directive or directives.
The #COBALT tag should be right after the first line of the script.
qsub will continue to read directives until it reads the first line that 
does not contain the tag #COBALT.

The command line options will take precedence over the Cobalt script jobx
directives.

Note: An error will occur if in the script the --mode option 
is specified and it is not "script"

Example: 

  script.sh:
     
    #!bin/bash
    #COBALT -t 50 -n 30
    echo "SCRIPT JOB WITH DIRECTIVE"

  or

    #!bin/bash
    #COBALT -t 50 
    #COBALT -n 30
    echo "SCRIPT JOB WITH DIRECTIVE"

  qsub script.sh sargs 
  is equivalent to 
  qsub -t 50 -n 30 --mode script script.sh sargs

  qsub -t 75 script.sh sargs 
  is equivalent to 
  qsub -t 75 -n 30 --mode script script.sh sargs

  qsub -n 48 --mode script script.sh sargs 
  is equivalent to 
  qsub -t 75 -n 48 --mode script script.sh sargs

.SH "OPTIONS"
.TP
.B \-d \-\-debug
Turn on communication debugging
.TP
.B \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-A \-\-project [project]
Associate the job with the allocation for project project. This is
used to properly account for machine usage. 
.TP
.B \-\-attrs [<attr1=val1>:<attr2=val2>:...]
Set a list of attributes for a job that must be fulfilled for a job to run.  
For instance, --attrs location=[blocklocation] will force a job to run on a 
specific block.
.TP
.B \-\-cwd [directory]
Tell the job to use the specified directory as the default current working directory
while running the job.
.TP
.B \-\-debuglog [path to file] - (jobid expansion)
Send debugging information from cobalt to file specified.
.TP
.B \-\-dependencies [<jobid1>:<jobid2>:...]
Set the dependencies for the job being submitted.  This job won't run until 
all jobs in the dependency list have finished and exited with a status of 0.
.TP
.B \-\-disable_preboot
If --mode script is specified, this will cause Cobalt to run the job script without
booting the block first.  By default the script is executed after the requested block is booted.
.TP
.B \-e \-\-error [path to file] - (jobid expansion)
Send job stderr to file specified.
.TP
.B \-\-env [<var1=value1>:<var=value2>] - (jobid expansion)
Set environment variables that will be passed into the job's environment.
If the the characters ":" and "=" are within the value, right hand side (rhs) of a environment variable assignment,
then need to quote the entire string and precede any ":" and "=" characters that are in the rhs with "\\". 
runtime. 

If multiple \-\-env options are specified then qsub will take the union.

e.g. qsub ... --env "A=B:C=x\\=1\\:y\\=2\\:z\\=3:D=ABC" ... --env "MY_NAME=cobalt" ...

A would equal B

C would equal x=1:y=2:z=3

D would equal ABC

and

MY_NAME would be equal to "cobalt"

.TP
.B \-\-jobname [job name] - (jobid expansion)
Sets Jobname. If this option is not provided then Jobname will be set to whatever -o option specified.'
.TP
.B \-\-geometry <geometry_spec>
Sets the geometry of the block that the jobs should run on.  This is in
the form of AxBxCxDxE.  This must be the geometry of the compute block.
This may not be a fraction of a block.
.TP
.B \-h \-\-held
Submit job and immediately place it in the user hold state
.TP
.B \-i \-\-inputfile [path to file]
Send file to job's stdin.
.TP
.B \-I \-\-interactive
Run an interactive command. During job execution, the current tty will
be connected with the running job.
To run interactive job you can use -I, --interactive or --mode interactive.
Running qsub in "interactive" mode will create a shell for the user while running
qsub.  If the user CTRL-D or performs an "exit" command from the shell then Cobalt
will terminate the "interactive" job.

.TP 
.B \-M \-\-notify [email address]
Send an email notification at the start and stop of the job to the
specified email address.
.TP
.B \-n \-\-nodecount [nodecount]
Specifies the node count for a job
.TP
.B \-o \-\-output [path to file] - (jobid expansion)
Send job stdout to file specified.
.TP
.B \-O \-\-outputprefix [output prefix] - (jobid expansion)
Use the specified prefix for both .output, .error and debuglog files. This will only take effect if \-o or \-e or \-\-debuglog not supplied. 
.TP
.B \-\-proccount [process count]
Specify the number of processes to start.
.TP
.B \-q \-\-queue [queue] 
Submit the job to the specified queue.
.TP
.B \-\-run_users \-\-user_list [<user1>:<user2>:...]
Sets a colon-separated list of users for the job being submitted.  All users
in this list will be able to execute cobalt commands to control the job. The 
submitting user is always able to run commands on a submitted job.
.TP
.B \-\-run_project
Appends the list of users in a project to the list of users authorized to run
commands on the submitted job. 
.TP
.B \-u \-\-umask
set umask: octal number default(022)
.TP
.B \-t \-\-time [wallclock time] 
Specify the runtime for a job. If the job runs over this limit, it will be 
killed. The time may be specified as eitehr an integer number of minutes or a 
colon-delimited value of the format: HH:MM:SS. Enter 0 to get the max allowed 
walltime.
.TP


.SH "BLUEGENE OPTIONS"
.TP 
The following commands are only valid on IBM BlueGene architecture platforms.
.TP
.B \-\-kernel [kernel profile]
Run the job with the specified kernel profile.
.TP
.B \-K \-\-kerneloptions [kernel options]
Run the job with the specified kernel options (equivalent to 
.Impirun -kernel_options 
on Blue Gene/L).
.TP
.B \-\-ion_kernel [kernel profile]
Run the job with the specified IO Node kernel profile. (BlueGene/Q only)
.TP
.B \-\-ion_kerneloptions [kernel options]
Run the job with the specified IO Node kernel options.  (BlueGene/Q only)
.TP
.B \-\-mode [mode]
This is a platform-dependent setting that controls the execution mode of the 
submitted job and the division of cores and memory on the compute nodes.
Valid values are dependent on the specific BlueGene platform:

.TP
BG/L:  Valid values are "co" (coprocessor),"vn" (virtual node) and "script".
The default is "co".

.TP
BG/P:  Valid values are "smp", "dual", "vn" and "script".   "smp", "dual" and
"vn" correspond to the modes within mpirun. The default is "smp".

.TP
BG/Q:  Valid values are "c1","c2","c4","c8","c16","c32","c64","script" and
"interactive".  The c(n) modes correspond to the "ranks-per-node" flag of
runjob and governs the MPI ranks started per node on the compute nodes.  Node
memory is divided evenly between all processes.  64/n threads will be available
to each MPI process as well. If this option is not set the default is "c1".

On all platforms, the "script" mode causes the job to be routed through the
script-forker, and will allow the execution of script jobs.  On linux cluster
platforms, all jobs are run as script jobs.

.SH "NOTE"
The only thing printed to STDOUT is the jobid, any other error or informational messages are printed to STDERR.
.SS "Cray Systems"
On Cray systems, the "location" attribute may be specified by a comma-delimited
list of node ids.  Runs of node ids may be compacted to a hyphenated, inclusive
pair, i.e. 1-4 would expand to 1, 2, 3, 4.  All nodes specified in location must
exist on the system.  This list format is compatible with the values returned by
Cray's
.BRcnselect (1)
command.
.SH "SEE ALSO"
.BR qstat (1),
.BR qdel (1),
.BR qhold (1),
.BR qrls (1),
.BR qalter (1),
.BR qmove (1),
.BR cqm (1),
.BR cqwait (1)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 0.99.42 ==
 * Update to database writer to handle attrs fields that are
   improperly formatted.

== Changes to 0.99.41 ==
 * Minor patch to clean up an issue where the user was not being properly
   cleared from BlueGene blocks at the end of a run.

== Changes to 0.99.40 ==
 * A nokillpg attribute may be added to a job to cause kill instead of killpg
   to be used for job termination so that a signal handler in a script may
   handle signal propigation itself.
 * Fixed a bug where a job requesting a location could cause Cobalt to drain
   for a job that would require the use of offline hardware
 * Added a check to the job validator where a job is rejected if the job size
   and block geometry have a mismatch.

== Changes to 0.99.39 ==
 * Session information added to job submission information.  This also includes
   submitting host information (Cobalt Ticket #4315)
 * A fix for a bug that can cause "phantom drains" to occur on lightly loaded
   cluster systems. (Cobalt Ticket #4319)

== Changes to 0.99.38 ==
 * Added numerous dependency changes.  Notably, jobs in dep_hold states will no
   longer accrue eligible wait time or score.  Previously, dep_hold jobs would
   accrue eligible wait times unlike admin_hold or user_hold states.
   (Cobalt Ticket #4308)
 * Two display fixes for cluster systems that allow multiple node-name prefixes
   to be handled correctly by the nodelist merging code  and to correctly
   compact names if leading 0's are used in location numbers like
   "foo001.machine". (Cobalt Ticket #4313 and #4277)
 * Corrected two race conditions in the BlueGene/Q system components where a
   location with offline hardware could be selected for draining on a
   scheduling pass.  Though Cobalt would never try to actually run on the
   offline block, It could end up draining resources needlessly, and impact
   machine usage.  The first could happen at any time, the second only occurred
   at the end of booting a block. (Cobalt Ticket #4311)

== Changes to 0.99.37 ==
 * Fixed issue where a reservation could hit it's activation check while it is being
   deactivated, causing those events should be reversed.  Released reservations can't
   activate. (Cobalt Ticket #4289)
 * COBALT_JOBSIZE and COBALT_PARTSIZE have been added to interactive jobs.
   (Cobalt Ticket #4294)
 * Job dependencies require the same project to gain score from eachother.
   (Cobalt Ticket #4293)
 * prologue_helper.py now handles "=" in arguments correctly. (Cobalt Ticket #4292)

== Changes to 0.99.36 ==
 * Fix for issues in get-bootable-blocks and with block booting.  Block names are now
   properly sorted.  A race condition between boot-block and update_block_state has
   been fixed.  The return statuses for boot-block have also been corrected for failed
   boots.  (Cobalt Ticket #4282 and #4280)

== Changes to 0.99.35 ==
 * Fixed an issue where a block being removed in the BGQ control system between block
   detection and the block being added for tracking could result in blocks being
   erroniously marked as idle. (Cobalt Ticket #4268)

== Changes to 0.99.34 ==
 * Fixed an issue where Cobalt could have a memory leak if boot-block.py was
   terminated early the boot tracking object may never be cleaned up.
 * Numerous enhancements and fixes to handle multiple overlaping queues in the
   system with respect to draining and backfilling. (Cobalt Ticket #663)
 * Fixed some typos in the documentation for qsub.
 * COBALT_RESID is now appropriately set in interactive mode jobs (Cobalt Ticket #703)

== Changes to 0.99.33 ==
 * Fixed issue where adding blocks while the system component is running can
   cause the system component's block update loop to fail. (Cobalt Ticket #695)
 * Preliminary fix for a racecondition that can cause subblock jobs to terminate
   early (Cobalt Ticket #690)
 * Added T-Minus to showres (Cobalt Ticket #692)
 * Fixed an error in qalter where attemtpts to change the users on a running job
   caused qalter to fail. (Cobalt Ticket #694)

== Changes to 0.99.32 ==
 * Cluster systems now drain for larger jobs, and will backfill into the
   draining hardware. (Cobalt Ticket #663)
 * Fixed a bug in the prologue_helper where a silent failure on sending the
   hostfile to nodes could occur silently (Cobalt Ticket #686)

== Changes to 0.99.31 == 
 * Interactive mode jobs (qsub -I) are now supported on the BlueGene/Q.
   (Cobalt Ticket #107)

src/clients/POSIX/qsub.py
#!/usr/bin/env python
"""
Submit jobs to the queue manager for execution.

Usage: %prog --help
Usage: %prog [options] <executable> [<excutable options>]

Refer to man pages for JOBID EXPANSION and SCRIPT JOB DIRECTIVES.

version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

Option with no values:

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'-v','--verbose',dest='verbose',help='not used',action='store_true'
'-h','--held',dest='user_hold',help='hold this job once submitted',action='store_true'
'--preemptable',dest='preemptable',help='make this job preemptable',action='store_true'
'--run_project',dest='run_project',help='set run project flag for this job',action='store_true'
'--disable_preboot',dest='script_preboot',help='disable script preboot',action='store_false'
'-I','--interactive',help='run qsub in interactive mode',callback=cb_interactive

Option with values:

'-n','--nodecount',dest='nodes',type='int',help='set job node count',callback=cb_nodes
'--proccount',dest='procs',type='int',help='set job proc count',callback=cb_gtzero
'-A','--project',dest='project',type='string',help='set project name'
'--cwd',dest='cwd',type='string',help='set current working directory'
'-q','--queue',dest='queue',type='string',help='set queue name'
'-M','--notify',dest='notify',type='string',help='set notification email address'
'--env',dest='envs',type='string', help='Set env variables. Refer to man pages for more detail information.', callback=cb_env
'-t','--time',dest='walltime',type='string',help='set walltime (minutes or HH:MM:SS). For max walltime enter 0.',callback=cb_time
'-u','--umask',dest='umask',type='string',help='set umask: octal number default(022)',callback=cb_umask
'-O','--outputprefix',dest='outputprefix',type='string',help='output prefix for error,output or debuglog files',callback=cb_path
'-e','--error',dest='errorpath',type='string',help='set error file path',callback=cb_path
'-o','--output',dest='outputpath',type='string',help='set output file path',callback=cb_path
'-i','--inputfile',dest='inputfile',type='string',help='set input file',callback=cb_path
'--debuglog',dest='cobalt_log_file',type='string',help='set debug log path file',callback=cb_path
'--dependencies',dest='all_dependencies',type='string',help='set job dependencies (jobid1:jobid2:...:jobidN)',callback=cb_dep
'--attrs',dest='attrs',type='string',help='set attributes (attr1=val1:attr2=val2:...:attrN=valN)',callback=cb_attrs
'--user_list','--run_users',dest='user_list',type='string',help='set user list (user1:user2:...:userN)',callback=cb_user_list
'--jobname',dest='jobname',type='string', /
   help='Sets Jobname. If this option is not provided then Jobname will be set to whatever -o option specified.'

The following options are only valid on IBM BlueGene architecture platforms:

'--kernel',dest='kernel',type='string',help='set a compute node kernel profile'
'-K','--kerneloptions',dest='kerneloptions',type='string',help='set compute node kernel options'
'--ion_kernel',dest='ion_kernel',type='string',help='set an IO node kernel profile'
'--ion_kerneloptions',dest='ion_kerneloptions',type='string',help='set IO node kernel options'
'--mode', dest='mode', type='string', help='select system mode',callback=cb_mode
'--geometry', dest='geometry', type='string', help='set geometry (AxBxCxDxE)',callback=cb_geometry

"""
import logging
import time
import string
import os
import sys
import signal
import socket
from Cobalt import client_utils
from Cobalt.client_utils import \
    cb_debug, cb_env, cb_nodes, cb_time, cb_umask, cb_path, cb_dep, \
    cb_attrs, cb_user_list, cb_geometry, cb_gtzero, cb_mode, cb_interactive
from Cobalt.arg_parser import ArgParse
from Cobalt.Util import get_config_option, init_cobalt_config, sleep
from Cobalt.Proxy import ComponentProxy
import xmlrpclib
import subprocess

__revision__ = '$Revision: 559 $'
__version__  = '$Version$'

#init cobalt config file for setting default kernels.
init_cobalt_config()
SYSMGR           = client_utils.SYSMGR
QUEMGR           = client_utils.QUEMGR
CN_DEFAULT_KERNEL  = get_config_option('bgsystem', 'cn_default_kernel', 'default')
ION_DEFAULT_KERNEL = get_config_option('bgsystem', 'ion_default_kernel', 'default')

def on_interrupt(sig, func=None):
    """
    Interrupt Handler to cleanup the interactive job if the user interrupts
    Will contain two static variables: count and exit.
    'count' will keep track how many interruptions happened and
    'exit' flags whether we completely exit once the interrupt occurs.
    """
    on_interrupt.count += 1
    if on_interrupt.exit:
        sys.exit(1)

# Initializing on_interrupt static variables
on_interrupt.count = 0
on_interrupt.exit  = True

def exit_on_interrupt():
    """
    Set to exit when interrupt occurs
    """
    on_interrupt.exit = True
    if on_interrupt.count > 0:
        sys.exit(1)

def not_exit_on_interrupt():
    """
    Set to not exit on interrupt
    """
    on_interrupt.exit = False

# Reset sigint and sigterm interrupt handlers to deal with interactive failures
signal.signal(signal.SIGINT, on_interrupt)
signal.signal(signal.SIGTERM, on_interrupt)
signal.signal(signal.SIGQUIT, on_interrupt)
signal.signal(signal.SIGABRT, on_interrupt)
signal.signal(signal.SIGXCPU, on_interrupt)
signal.signal(signal.SIGPIPE, on_interrupt)

def exit_interactive_job(deljob, jobid, user):
    """
    Exit job normally or delete job
    """
    not_exit_on_interrupt()
    # If no jobid assigned yet return
    if not jobid:
        return
    if deljob:
        job_to_del = [{'tag':'job', 'jobid':jobid, 'user':user}]
        client_utils.logger.info("Deleting interactive job %s", str(jobid))
        client_utils.component_call(QUEMGR, False, 'del_jobs', (job_to_del, False, user))
    else:
        client_utils.logger.info("Exiting interactive job %d", int(jobid))
        client_utils.component_call(SYSMGR, False, 'interactive_job_complete', (jobid,))

def validate_args(parser, spec):
    """
    If the argument is a script job it will validate it, and get the Cobalt directives
    """

    #an executable cannot be specified for interactive jobs
    if parser.options.mode == 'interactive' and (not parser.no_args()):
        client_utils.logger.error("An executable may not be specified if using the interactive option.")
        sys.exit(1)
    elif parser.options.mode == 'interactive':
        #Bypass the rest of the checks for interactive jobs.
        return

    # if no excecutable specified then flag it an exit
    if parser.no_args():
        client_utils.print_usage(parser, "No executable or script specified")
        sys.exit(1)

    # Check if it is a valid executable/file
    cmd = parser.args[0].replace(' ','')
    if cmd[0] != '/':
        cmd = spec['cwd'] + '/' + cmd

    if not os.path.isfile(cmd):
        client_utils.logger.error("command %s not found, or is not a file" % cmd)
        sys.exit(1)

    if not os.access(cmd, os.X_OK | os.R_OK):
        client_utils.logger.error("command %s is not executable" % cmd)
        sys.exit(1)

    # if there are any arguments for the specified command store them in spec
    spec['command'] = cmd
    if len(parser.args) > 1:
        spec['args'] = parser.args[1:]
    else:
        spec['args'] = []

    tag         = '#COBALT '
    len_tag     = len(tag)
    fd          = open(cmd, 'r')
    line        = fd.readline()
    reparse    = False
    if line[0:2] == '#!':
        line                = fd.readline()
        new_argv            = []
        while len(line) > len_tag:
            if line[:len_tag] != tag:
                break
            if new_argv == []:
                new_argv = ['--mode','script']
                reparse   = True
            new_argv += line[len_tag:].split()
            line      = fd.readline()
        sys.argv = [sys.argv[0]] + new_argv + sys.argv[1:]
    fd.close()

    # check for multiple --env options specified
    if sys.argv.count('--en') + sys.argv.count('--env') > 1:
        # consolidate the --env options (this will update sys.argv)
        env_union()
        reparse = True

    return reparse

def validate_options(parser, opt_count):
    """
    Validate qsub arguments
    """

    # Check if any required option entered
    if opt_count == 0:
        client_utils.print_usage(parser, "No required options provided")
        sys.exit(1)

    # If no time supplied flag it and exit
    if parser.options.walltime == None:
        client_utils.logger.error("'time' not provided")
        sys.exit(1)

    # If no nodecount give then flag it an exit
    if parser.options.nodes == None:
        client_utils.logger.error("'nodecount' not provided")
        sys.exit(1)

def update_outputprefix(parser,spec):
    """
    Update the the appropriate paths with the outputprefix path
    """
    # If the paths for the error log, output log, or the debuglog are not provided
    # then update them with what is provided in outputprefix.
    if parser.options.outputprefix != None:
        # pop the value for outputrefix
        op = spec.pop('outputprefix')

        # if error path, cobalt log path or output log path not provide then update them wiht outputprefix
        if parser.options.errorpath == None:
            spec['errorpath'] = op + ".error"
        if parser.options.cobalt_log_file == None:
            spec['cobalt_log_file'] = op + ".cobaltlog"
        if parser.options.outputpath == None:
            spec['outputpath'] = op + ".output"

def update_paths(spec):
    """
    This functiojn will update all the paths in spec that need the current working directory.
    """
    for key in spec:
        value = spec[key]
        # if path needs current working directory then
        if type(value) == type(""):
            if value.find(client_utils.CWD_TAG) != -1:
                if 'cwd' in spec:
                    _cwd = spec['cwd']
                else:
                    _cwd = client_utils.getcwd()
                _path = spec[key].replace(client_utils.CWD_TAG,_cwd)
                # validate the path
                if not os.path.isdir(os.path.dirname(_path)):
                    client_utils.logger.error("directory %s does not exist" % _path)
                    sys.exit(1)
                spec[key] = _path

def check_inputfile(parser, spec):
    """
    Verify the input file is an actual file
    """
    if parser.options.inputfile != None:
        if parser.options.mode == 'interactive':
            client_utils.logger.error("Cannot specify an input file for interactive jobs.")
            sys.exit(1)
        inputfile = spec['inputfile']
        if not os.path.isfile(inputfile):
            client_utils.logger.error("file %s not found, or is not a file" % inputfile)
            sys.exit(1)

def update_spec(parser, opts, spec, opt2spec):
    """
    This function will update the appropriate spec values with the opts values
    """
    # Get the key validated values into spec dictionary
    for opt in ['mode', 'proccount', 'nodecount']:
        spec[opt2spec[opt]] = opts[opt]

    # Hack until the Cluster Systems get re-written.
    if parser.options.mode == 'interactive' and 'command' in opts and 'args' in opts:
        spec['command'] = opts['command']
        spec['args']    = opts['args']

    #stamp the tty info, if it exists into the spec.
    spec['ttysession'] = fetch_tty_session()
    spec['submithost'] = socket.gethostname()

def fetch_tty_session():
    '''Grab the tty session stdout is attached to.  This will return None
    if stdout is not going to a terminal.

    '''
    ttyname = None
    try:
        stdoutfh = sys.stdout.fileno()
        ttyname = os.ttyname(stdoutfh)
    except (OSError, IOError):
        client_utils.logger.debug("fd %d not associated with a terminal device", stdoutfh)
    return ttyname

def logjob(jobid, spec, log_to_console, ttyname=None):
    """
    log job info
    """
    # log jobid to stdout
    if jobid:
        if log_to_console:
            client_utils.logger.info(jobid)
        if spec.has_key('cobalt_log_file'):
            filename = spec['cobalt_log_file']
            template = string.Template(filename)
            filename = template.safe_substitute(jobid=jobid)
        else:
            filename = "%s/%s.cobaltlog" % (spec['outputdir'], jobid)
        try:
            with open(filename, "a") as cobalt_log_file:
                print >> cobalt_log_file, "Jobid: %s" % jobid
                print >> cobalt_log_file, "qsub %s" % (" ".join(sys.argv[1:]))
                print >> cobalt_log_file, ("%s submitted with cwd set to: %s" %
                    (client_utils.sec_to_str(time.time()), spec['cwd']))
                if ttyname is not None:
                    print >> cobalt_log_file, ("jobid %d submitted from terminal %s" %
                        (jobid, ttyname))
        except IOError:
            client_utils.logger.error("WARNING: failed to create cobalt log file at: %s: %s", filename, exc_info=True)
    else:
        client_utils.logger.error("failed to create the job.  Maybe a queue isn't there?")

def env_union():
    """
    Take all env options and their values in sys.argv and make the union of them to create one consolidated --env.
    """
    try:
        ndx = 0
        new_args    = []
        env_values  = []
        env_val_ndx = -1
        skip_next   = False
        for arg in sys.argv:
            if arg == '--en' or arg == '--env':
                env_values.append(sys.argv[ndx+1])
                if env_val_ndx == -1:
                    env_val_ndx = ndx+1
                    new_args.append('--env')
                    new_args.append('<PLACE HOLDER>')
                skip_next = True
            elif not skip_next:
                new_args.append(sys.argv[ndx])
            else:
                skip_next = False
            ndx += 1
        new_args[env_val_ndx] = ':'.join(env_values)
        sys.argv = new_args
    except Exception, e:
        client_utils.logger.error( "No values specified or invalid usage of --env option: %s --> %s", str(sys.argv), e)
        sys.exit(1)

def parse_options(parser, spec, opts, opt2spec, def_spec):
    """
    Will initialize the specs and then parse the command line options
    """
    opts.clear()
    for item in def_spec:
        spec[item] = def_spec[item]

    parser.parse_it() # parse the command line
    opt_count               = client_utils.get_options(spec, opts, opt2spec, parser)
    opts['disable_preboot'] = not spec['script_preboot']
    return opt_count

def fetch_pgid(user, jobid, loc, pgid=None):
    '''fetch and set pgid for user shell.  Needed for cray systems'''
    if client_utils.component_call(SYSMGR, False, 'get_implementation', ()) == 'alps_system':
        #Cray is apparently using the session id for interactive jobs.
        spec = [{'user': user, 'jobid': jobid, 'pgid': pgid, 'location':loc}]
        if not client_utils.component_call(SYSMGR, False, 'confirm_alps_reservation', (spec)):
            client_utils.logger.error('Unable to confirm ALPS reservation.  Exiting.')
            sys.exit(1)
    return

def exec_user_shell(user, jobid, loc):
    '''Execute shell for user for interactive jobs.  Uses the user's defined
    SHELL.  Will also send the pgid for cray systems so that aprun will work
    from within the shell.

    Wait until termination.

    '''
    pgid = os.getppid()
    proc = subprocess.Popen([os.environ['SHELL']], shell=True,
            preexec_fn=(lambda: fetch_pgid(user, jobid, loc, pgid=pgid)))
    os.waitpid(proc.pid, 0)

def run_interactive_job(jobid, user, disable_preboot, nodes, procs):
    """
    This will create the shell or ssh session for user
    """
    not_exit_on_interrupt()
    # save whether we are running on a cluster system
    impl =  client_utils.component_call(SYSMGR, False, 'get_implementation', ())
    exit_on_interrupt()
    deljob = True if impl == "cluster_system" else False

    def start_session(loc, resid, nodes, procs):
        """
        start ssh or shell session
        """
        # Create necesary env vars
        os.putenv("COBALT_NODEFILE", "/var/tmp/cobalt.%s" % (jobid))
        os.putenv("COBALT_JOBID", "%s" % (jobid))
        if resid:
            os.putenv("COBALT_RESID", "%s" % (resid))
        os.putenv("COBALT_PARTNAME", loc)
        os.putenv("COBALT_BLOCKNAME", loc)
        os.putenv("COBALT_JOBSIZE", str(procs))
        os.putenv("COBALT_BLOCKSIZE",str(nodes))
        os.putenv("COBALT_PARTSIZE", str(nodes))
        client_utils.logger.info("Opening interactive session to %s", loc)
        if deljob:
            os.system("/usr/bin/ssh -o \"SendEnv COBALT_NODEFILE COBALT_JOBID\" %s" % (loc))
        if impl == 'alps_system':
            exec_user_shell(user, jobid, loc)
        else:
            os.system(os.environ['SHELL'])

    # Wait for job to start
    query = [{'tag':'job', 'jobid':jobid, 'location':'*', 'state':"*",
        'resid':"*"}]
    client_utils.logger.info("Wait for job %s to start...", str(jobid))

    while True:
        # If we get a ssl timeout error or component lookup error try again
        try:
            not_exit_on_interrupt()
            response =  client_utils.component_call(QUEMGR, False, 'get_jobs', (query, ), False)
            exit_on_interrupt()
            # if jobid not found flag an error and exit
            if not response:
                client_utils.logger.error("Jobid %s not found after submission", str(jobid))
                sys.exit()
        except (xmlrpclib.Fault, ComponentProxy) as fault:
            # This can happen if the component is down so try again
            client_utils.logger.error('Error getting job info: %s. Try again', fault)
            sleep(2)
        state    = response[0]['state']
        location = response[0]['location']
        resid    = response[0]['resid']
        if state == 'running' and location:
            start_session(location[0], resid, nodes, procs)
            break
        client_utils.logger.debug('Current State "%s" for job %s', str(state), str(jobid))
        sleep(2)

    return deljob

def run_job(parser, user, spec, opts):
    """
    run the job
    """
    jobid        = None
    deljob       = True
    exc_occurred = False
    try:
        not_exit_on_interrupt()
        jobs  =  client_utils.component_call(QUEMGR, False, 'add_jobs',([spec],), False)
        jobid = jobs[0]['jobid']
        exit_on_interrupt()

        if parser.options.envs:
            client_utils.logger.debug("Environment Vars: %s", parser.options.envs)

        # If this is an interactive job, wait for it to start, then start user shell
        if parser.options.mode == 'interactive':
            logjob(jobid, spec, False, spec['ttysession'])
            deljob = run_interactive_job(jobid, user,  opts['disable_preboot'], opts['nodecount'], opts['proccount'])
        else:
            logjob(jobid, spec, True, spec['ttysession'])
    except Exception, e:
        client_utils.logger.error(e)
        exc_occurred = True
    finally:
        if parser.options.mode == 'interactive':
            exit_interactive_job(deljob, jobid, user)
        if exc_occurred: 
            sys.exit(1)

def main():
    """
    qsub main function.
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    spec     = {} # map of destination option strings and parsed values
    opts     = {} # old map
    opt2spec = {}
    def_spec = {}

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args (tuple) >
        ( cb_debug        , () ),
        ( cb_interactive  , () ),
        ( cb_env          , (opts,) ),
        ( cb_nodes        , (False,) ), # return string
        ( cb_gtzero       , (False,) ), # return string
        ( cb_time         , (False, False, False) ), # no delta time, minutes, return string
        ( cb_umask        , () ),
        ( cb_path         , (opts, True) ), # use CWD
        ( cb_dep          , () ),
        ( cb_attrs        , () ),
        ( cb_mode         , () ),
        ( cb_user_list    , (opts,) ),
        ( cb_geometry     , (opts,) )]

    # Get the version information
    opt_def =  __doc__.replace('__revision__', __revision__)
    opt_def =  opt_def.replace('__version__', __version__)

    user = client_utils.getuid()

    def_spec['tag']            = 'job'
    def_spec['user']           = user
    def_spec['outputdir']      = client_utils.CWD_TAG
    def_spec['jobid']          = '*'
    def_spec['path']           = client_utils.getpath()
    def_spec['mode']           = False
    def_spec['cwd']            = client_utils.getcwd()
    def_spec['kernel']         = CN_DEFAULT_KERNEL
    def_spec['ion_kernel']     = ION_DEFAULT_KERNEL
    def_spec['queue']          = 'default'
    def_spec['umask']          = 022
    def_spec['run_project']    = False
    def_spec['user_list']      = [user]
    def_spec['procs']          = False
    def_spec['script_preboot'] = True

    parser    = ArgParse(opt_def, callbacks)
    opt_count = parse_options(parser, spec, opts, opt2spec, def_spec)
    reparse  = validate_args(parser, spec)

    if reparse:
        # re-parse with new sys.argv
        # note: the first parse is necessary to make sure that
        #       the env syntax is correct for every --env option provided
        #       If not parsed prior to the union then the union could result
        #       in a valid syntax, but it would not be what the user would want.
        opt_count = parse_options(parser, spec, opts, opt2spec, def_spec)

    client_utils.setumask(spec['umask'])
    validate_options(parser, opt_count)
    update_outputprefix(parser, spec)
    update_paths(spec)
    check_inputfile(parser, spec)

    not_exit_on_interrupt()
    opts = client_utils.component_call(SYSMGR, False, 'validate_job',(opts,))
    exit_on_interrupt()

    filters = client_utils.get_filters()
    client_utils.process_filters(filters, spec)
    update_spec(parser, opts, spec, opt2spec)

    run_job(parser, user, spec, opts)

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, exc:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", exc)
        sys.exit(1)
